async-nats/src/jetstream/object_store/mod.rs

```diff
@@ -539,63 +539,68 @@ impl tokio::io::AsyncRead for Object<'_> {
+                            }
+                            self.remaining_bytes
+                                .extend_from_slice(&message.payload[len..]);
+
+                            let info = message.info().map_err(|err| {
+                                std::io::Error::new(
+                                    std::io::ErrorKind::Other,
+                                    format!("error from JetStream subscription: {err}"),
+                                )
+                            })?;
+                            if info.pending == 0 {
+                                let digest = self.digest.take().map(|context| context.finish());
+                                if let Some(digest) = digest {
+                                    if self
+                                        .info
+                                        .digest
+                                        .as_ref()
+                                        .map(|digest_self| {
+                                            format!("SHA-256={}", URL_SAFE.encode(digest))
+                                                != *digest_self
+                                        })
+                                        .unwrap_or(false)
+                                    {
+                                        return Poll::Ready(Err(io::Error::new(
+                                            ErrorKind::InvalidData,
+                                            "wrong digest",
+                                        )));
+                                    }
+                                } else {
                                     return Poll::Ready(Err(io::Error::new(
                                         ErrorKind::InvalidData,
-                                        "wrong digest",
+                                        "digest should be Some",
                                     )));
                                 }
-                            } else {
-                                return Poll::Ready(Err(io::Error::new(
-                                    ErrorKind::InvalidData,
-                                    "digest should be Some",
-                                )));
+                                self.has_pending_messages = false;
+                                self.subscription = None;
                             }
-                            self.has_pending_messages = false;
+                            Poll::Ready(Ok(()))
                         }
-                        Poll::Ready(Ok(()))
-                    }
-                    None => Poll::Ready(Err(std::io::Error::new(
-                        std::io::ErrorKind::Other,
-                        "subscription ended before reading whole object",
-                    ))),
-                },
-                Poll::Pending => Poll::Pending,
+                        None => Poll::Ready(Err(std::io::Error::new(
```

Comment:
Should we gracefully try to recover from this?

Reply:
I think that if the ordered consumer underneath cannot recover, this should not try either.